shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -188,39 +269,65 @@ private void invokeHandler(
     }
   }
 
-  @Override
-  public void handleMessageFromDart(
-      @NonNull final String channel,
+  private void dispatchMessageToQueue(
+      @NonNull String channel,
+      @Nullable HandlerInfo handlerInfo,
       @Nullable ByteBuffer message,
-      final int replyId,
+      int replyId,
       long messageData) {
-    // Called from the ui thread.
-    Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
-    @Nullable final HandlerInfo handlerInfo = messageHandlers.get(channel);
-    @Nullable
     final DartMessengerTaskQueue taskQueue = (handlerInfo != null) ? handlerInfo.taskQueue : null;
     Runnable myRunnable =
         () -> {
           Trace.beginSection("DartMessenger#handleMessageFromDart on " + channel);
           try {
             invokeHandler(handlerInfo, message, replyId);
             if (message != null && message.isDirect()) {
-              // This ensures that if a user retains an instance to the ByteBuffer and it happens to
-              // be direct they will get a deterministic error.
+              // This ensures that if a user retains an instance to the ByteBuffer and it
+              // happens to be direct they will get a deterministic error.
               message.limit(0);
             }
           } finally {
+            Trace.endSection();
```

Comment:
Move this below `flutterJNI.cleanupMessageData`

`flutterJNI.cleanupMessageData` must be called in order to avoid memory leaks.  Calling it first avoids the potential risk of `Trace.endSection` throwing.